File: recursion.java
class Node {
    int data;
    Node left, right;
    
    // Конструктор узла
    public Node(int item) {
        this.data = item;
        this.left = null;
        this.right = null;
    }
}

public class BinaryTreeLeafCounter {

    // Метод для подсчета листьев в бинарном дереве
    static int countLeaves(Node root) {
        if (root == null)
            return 0;
        
        // Если узел является листом (нет детей слева и справа)
        if (root.left == null && root.right == null)
            return 1;
            
        // Рекурсивно считаем листья в левом и правом поддереве
        else
            return countLeaves(root.left) + countLeaves(root.right);
    }

    public static void main(String[] args) {
        /* Пример бинарного дерева:
              1
           /     \
          2       3
         / \     /
        4   5   6
               /
              7
        */
        
        // Создаем дерево вручную
        Node root = new Node(1); 
        root.left = new Node(2); 
        root.right = new Node(3); 
        root.left.left = new Node(4); 
        root.left.right = new Node(5); 
        root.right.left = new Node(6); 
        root.right.left.left = new Node(7); 
        
        // Вычисляем количество листьев
        System.out.println("Количество листьев в данном бинарном дереве: " + countLeaves(root));
    }
}

вывод:Количество листьев в данном бинарном дереве: 3

File: recursion.cpp
#include <iostream>
using namespace std;

struct TreeNode {
    int data;
    TreeNode* left;
    TreeNode* right;
    TreeNode(int val) : data(val), left(nullptr), right(nullptr) {}
};

int countLeaves(TreeNode* node) {
    if (node == nullptr) return 0;
    if (node->left == nullptr && node->right == nullptr) return 1;
    return countLeaves(node->left) + countLeaves(node->right);
}

int main() {
    TreeNode* root = new TreeNode(1);
    root->left = new TreeNode(2);
    root->right = new TreeNode(3);
    root->left->left = new TreeNode(4);
    root->left->right = new TreeNode(5);
    root->right->left = new TreeNode(6);
    root->right->right = new TreeNode(7);

    cout << "Количество листьев: " << countLeaves(root) << endl;  // Выведет: 4

    // Освобождение памяти
    delete root->left->left;
    delete root->left->right;
    delete root->right->left;
    delete root->right->right;
    delete root->left;
    delete root->right;
    delete root;

    return 0;
}

вывод: 
Количество листьев: 4
